testbed/tests/window/test_window.py

```diff
@@ -507,6 +689,130 @@ async def test_move_and_resize(second_window, second_window_probe):
+            # Switch from MAXIMIZED:
+            (WindowState.MAXIMIZED, WindowState.NORMAL),
+            (WindowState.MAXIMIZED, WindowState.MINIMIZED),
+            (WindowState.MAXIMIZED, WindowState.FULLSCREEN),
+            (WindowState.MAXIMIZED, WindowState.PRESENTATION),
+            (WindowState.MAXIMIZED, WindowState.MAXIMIZED),
+            # Switch from FULLSCREEN:
+            (WindowState.FULLSCREEN, WindowState.NORMAL),
+            (WindowState.FULLSCREEN, WindowState.MINIMIZED),
+            (WindowState.FULLSCREEN, WindowState.MAXIMIZED),
+            (WindowState.FULLSCREEN, WindowState.PRESENTATION),
+            (WindowState.FULLSCREEN, WindowState.FULLSCREEN),
+            # Switch from PRESENTATION:
+            (WindowState.PRESENTATION, WindowState.NORMAL),
+            (WindowState.PRESENTATION, WindowState.MINIMIZED),
+            (WindowState.PRESENTATION, WindowState.MAXIMIZED),
+            (WindowState.PRESENTATION, WindowState.FULLSCREEN),
+            (WindowState.PRESENTATION, WindowState.PRESENTATION),
+        ],
+    )
+    @pytest.mark.parametrize(
+        "intermediate_states",
+        [
+            (
+                WindowState.MINIMIZED,
+                WindowState.FULLSCREEN,
+                WindowState.PRESENTATION,
+                WindowState.MAXIMIZED,
+                WindowState.MINIMIZED,
+                WindowState.FULLSCREEN,
+            ),
+            (
+                WindowState.FULLSCREEN,
+                WindowState.MAXIMIZED,
+                WindowState.MINIMIZED,
+                WindowState.PRESENTATION,
+                WindowState.FULLSCREEN,
+                WindowState.MINIMIZED,
+            ),
+        ],
+    )
+    @pytest.mark.parametrize(
+        "second_window_class, second_window_kwargs",
+        [
+            (
+                toga.MainWindow,
+                dict(title="Secondary Window", position=(200, 150)),
+            )
+        ],
+    )
+    async def test_window_state_change_with_intermediate_states(
+        app,
+        app_probe,
+        initial_state,
+        final_state,
+        second_window,
+        second_window_probe,
+        intermediate_states,
+    ):
+        """Window state can be changed to another state while passing
+        through intermediate states with an expected OS delay."""
+        if (
+            WindowState.MINIMIZED in {initial_state, final_state}
+            and not second_window_probe.supports_minimize
+        ):
+            pytest.xfail(
+                "This backend doesn't reliably support minimized window state."
+            )
+        second_window.toolbar.add(app.cmd1)
+        second_window.content = toga.Box(style=Pack(background_color=CORNFLOWERBLUE))
+        second_window.show()
+
+        # Add delay to ensure windows are visible after animation.
+        await second_window_probe.wait_for_window("Secondary window is visible")
+
+        assert second_window.state == WindowState.NORMAL
```

Comment:
As above, this needs to be a check of the "true" state.

Reply:
I have added a new property on the window probe `window_probe.instantaneous_state` that reports the actual window state and not the pending state.